File: spark/src/main/scala/geotrellis/spark/io/AttributeCaching.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.io

import geotrellis.spark._
import com.github.blemale.scaffeine.Scaffeine
import geotrellis.spark.io.hadoop.conf.AttributeConfig

import scala.concurrent.duration._
import spray.json._
import spray.json.DefaultJsonProtocol._

trait AttributeCaching { self: AttributeStore =>
  @transient private lazy val cache =
    Scaffeine()
      .recordStats()
      .expireAfterWrite(AttributeConfig.caching.expirationMinutes.minutes)
      .maximumSize(AttributeConfig.caching.maxSize)
      .build[(LayerId, String), JsValue]

  def cacheRead[T: JsonFormat](layerId: LayerId, attributeName: String): T =
    if(AttributeConfig.caching.enabled)
      cache.get(layerId -> attributeName, { _ => read[JsValue](layerId, attributeName) }).convertTo[T]
    else
      read[JsValue](layerId, attributeName).convertTo[T]

  def cacheLayerType(layerId: LayerId, layerType: LayerType): LayerType =
    if (AttributeConfig.caching.enabled)
      cache.get(layerId -> "layerType", { _ => layerType.toJson }).convertTo[LayerType]
    else
      layerType

  def cacheWrite[T: JsonFormat](layerId: LayerId, attributeName: String, value: T): Unit = {
    if(AttributeConfig.caching.enabled) cache.put(layerId -> attributeName, value.toJson)
    write[T](layerId, attributeName, value)
  }

  def clearCache(): Unit = {
    if(AttributeConfig.caching.enabled) cache.invalidateAll()
  }

  def clearCache(id: LayerId): Unit = {
    if(AttributeConfig.caching.enabled) {
      val toInvalidate = cache.asMap.keys.filter(_._1 == id)
      cache.invalidateAll(toInvalidate)
    }
  }

  def clearCache(id: LayerId, attribute: String): Unit = {
    if(AttributeConfig.caching.enabled) cache.invalidate(id -> attribute)
  }
}

File: geotrellis-python/geotrellis/spark/io/AttributeCaching.py
class AttributeCaching(object):
    def __init__(self):
        self._cache = {}

    def cacheRead(self, attr_type, layer_id, attr_name):
        tup = (layer_id, attr_name)
        if tup in self._cache:
            return self._cache[tup]
        else:
            attr = self.read(attr_type, layer_id, attr_name)
            self._cache[tup] = attr
            return attr

    def cacheWrite(self, attr_type, layer_id, attr_name, value):
        tup = (layer_id, attr_name)
        self._cache[tup] = value
        self.write(attr_type, layer_id, attr_name, value)

    def clearCache(self, layer_id = None, attr_name = None):
        if layer_id is None:
            if attr_name is None:
                self._cache.clear()
            else:
                raise Exception("cannot clear attr when layer_id is unknown")
        else:
            if attr_name is None:
                for tup in filter(lambda each: each[0] == layer_id, self._cache):
                    del self._cache[tup]
            else:
                del self._cache[(layer_id, attr_name)]

